src/module.c

```diff
@@ -13489,6 +13512,153 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
     return ctx->dbid;
 }
 
+/* --------------------------------------------------------------------------
+ * ## Key API for Set type
+ *
+ * See also RM_ValueLength(), which returns the length of a set.
+ * -------------------------------------------------------------------------- */
+
+/* Add new elements into a set. If the key is an empty key open for writing,
+ * it is created with an empty set key in order to add the elements.
+ *
+ * * **key**: The set key to be added elements to.
+ * * **fmt**: A format specifier string for the command's arguments. Each
+ *   of the arguments should be specified by a valid type specification.
+ *
+ *     * `c` -- The argument is a pointer to a plain C string (null-terminated).
+ *     * `l` -- The argument is long long integer.
+ *     * `s` -- The argument is a RedisModuleString.
+ *     * `v` -- The argument(s) is a vector of RedisModuleString.
+ * * **...**: The actual arguments to the Redis command.
```

Comment:
i had two arguments in the past.
1. that maybe we want to let the caller pass plain C strings and so on.
2. that maybe we want to be consistent with the Hash API.

if i'm not missing anything, this solves the first issue, but instead of being consistent with the Hash API (e.g. `REDISMODULE_HASH_CFIELDS`), it is consistent with `RM_Replicate`.

@zuiderkwast @itamarhaber maybe you have a suggestion?

Reply:
The idea for the RedisModule_SetAdd and RedisModule_SetRem APIs are to write data locally. Replication could be done via RedisModule_Replicate() while implementing extend redis command which accessing set data by invoking RedisModule_SetAdd or RedisModule_SetRem

Reply:
OK. Let me try to optimize it.

Reply:
For RM_HashSet, it is different because there you set the field's *value* to the special `REDISMODULE_HASH_DELETE` to delete the field. We can't do the same with a set since an element is not a field-value pair.

I like the three functions RM_SetAdd and RM_SetRem RM_SetIsMember.

Regarding format flags, I think it's better to use an `int flags` (like RM_HashSet) because the string flags for RM_Call() are for multiple arguments. In RM_Call(), each argument has its own format specifier (like printf) but for these Set API functions, all elements are of the same type and we only want one type.

Reply:
So there will be three functions RM_SetAdd, RM_SetRem and RM_SetIsMember instead combined one. and `int flags` to determine the left arguments are RedisModuleString or  plain C strings.  Is it OK to you?

Reply:
I realized I have no idea while writing the simulator to invoke the RM_SetAdd after changing the RM_SetAdd as RM_SetAdd(RedisModuleKey *key, int flags, ...) because the elements number to be added is variable. e.g.

int set_add(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
    if (argc < 3) return RedisModule_WrongArity(ctx);
    RedisModule_AutoMemory(ctx);
    int keymode = REDISMODULE_READ | REDISMODULE_WRITE;
    RedisModuleKey *key = RedisModule_OpenKey(ctx, argv[1], keymode);
    RedisModule_ReplyWithLongLong(ctx, RedisModule_SetAdd(key, REDISMODULE_SET_NONE, **??????**);
    RedisModule_CloseKey(key);
    return REDISMODULE_OK;
}

But with fmt. it is easy to call. e.g.

int set_add(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
    if (argc < 3) return RedisModule_WrongArity(ctx);
    RedisModule_AutoMemory(ctx);
    int keymode = REDISMODULE_READ | REDISMODULE_WRITE;
    RedisModuleKey *key = RedisModule_OpenKey(ctx, argv[1], keymode);
    RedisModule_ReplyWithLongLong(ctx, RedisModule_SetAdd(key, "v", &argv[2], argc-2));
    RedisModule_CloseKey(key);
    return REDISMODULE_OK;
}

Reply:
Have you seen how it is done in [`RM_HashSet`](https://redis.io/docs/reference/modules/modules-api-ref/#RedisModule_HashSet)? The argument list is NULL terminated.

For format string, if you add multiple elements, would you repeat the format for each element? Like `RedisModule_SetAdd(key, "ssss", s1, s2, s3, s4)`?

Reply:
The problem is not at function implementation. it is at the caller which invokes the function. To be honest, RM_HashSet has limitation for caller with the number of arguments must be known while invoking the function if you look at the simulator. 
https://github.com/redis/redis/blob/2495b90a647f9f9987202efd29647f81f217b8ad/tests/modules/hash.c#L21

    _if (argc < 5 || argc % 2 == 0 || **argc > 11**)
        return RedisModule_WrongArity(ctx);
   ......
       **if (argc == 5)** {
        result = RedisModule_HashSet(key, flags,
                                     argv[3], value_or_delete(argv[4]),
                                     NULL);
    } **else if (argc == 7)** {
        result = RedisModule_HashSet(key, flags,
                                     argv[3], value_or_delete(argv[4]),
                                     argv[5], value_or_delete(argv[6]),
                                     NULL);
    } **else if (argc == 9)** {
        result = RedisModule_HashSet(key, flags,
                                     argv[3], value_or_delete(argv[4]),
                                     argv[5], value_or_delete(argv[6]),
                                     argv[7], value_or_delete(argv[8]),
                                     NULL);
    } **else if (argc == 11)** {
        result = RedisModule_HashSet(key, flags,
                                     argv[3], value_or_delete(argv[4]),
                                     argv[5], value_or_delete(argv[6]),
                                     argv[7], value_or_delete(argv[8]),
                                     argv[9], value_or_delete(argv[10]),
                                     NULL);
    } else {
        return RedisModule_ReplyWithError(ctx, "ERR too many fields");
    }_

It is not friendly. The number of elements to be added/removed to set at key shall not be limited if referring to SADD command.

For example the scenario while implementing one redis extend command:
1.  set multiple(unknown number) pairs of string type key/value data.
2.  add the keys to one set at key for sake of finding keys performance from the set at key instead of from whole redis.

You'd better not limit the number of pairs key/value from the extend command..

The benefit with fmt is the "v" can tell the the function next argument is RedisModuleString type vector with number of elements is in next argument following.

Reply:
You are right, `RM_HashSet` is not perfect. :-) I was just saying what we need to do if we want the syntax to match `RM_HashSet`.

Personally, I think these functions don't need to be variadic at all. If we wouldn't already have variadic HashSet, I would prefer a simple API `RM_HashSet(key, flags, field, value)` and let the module loop over its data and call this functions many times. But now we already have a variadict HashSet so maybe we should have SetAdd variadic too.

But we can allow a in flag corresponding to "v" as well, so it is possible to do something like

```C
RedisModule_SetAdd(key, REDISMODULE_SET_ARGV_ARGC, argv, argc)
```

or possibly, if we want to allow a variadic number of vectors too, expect a NULL-terminated list of argv-argc pairs

```C
RedisModule_SetAdd(key, REDISMODULE_SET_ARGV_ARGC, argv1, argc1, argv2, argc2, NULL)
```

But maybe this would make the API too complex; too hard to understand for users. It is really just as efficient to call it many times and easy to understand and write code like this:

```C
for (i = 0; i < argc; i++) {
    num_added += RedisModule_SetAdd(key, REDISMODULE_SET_NONE, argv[i], NULL);
}
```

That's some possibilities. I think it's up to the core team (@oranagra) to make the final decisions.